pkg/events/system.go

```diff
@@ -0,0 +1,178 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package events
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	errors2 "github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"regexp"
+)
+
+var eventFileRegex *regexp.Regexp = nil
+
+const eventFileFormat = "\\d{17}-([a-zA-Z]+)\\.json"
+
+func init() {
+	r, err := regexp.Compile(eventFileFormat)
+	if err != nil {
+		panic(err)
+	}
+	eventFileRegex = r
+}
+
+type EventSystem interface {
+	// RegisterEventHandler registers an event handler for the given type, which will be called when the an event of this
+	// type is received.
+	RegisterEventHandler(eventType reflect.Type, handler EventHandler)
+	RegisteredEventTypes() []reflect.Type
+	ProcessEvent(event interface{}) error
+	PublishEvent(event interface{}) error
+	LoadAndApplyEvents(location string) error
+}
+
+type EventHandler func(interface{}) error
+
+type eventSystem struct {
+	eventHandlers map[reflect.Type]EventHandler
+	// lastLoadedEvent contains the file name of the last event that was loaded. It is used
+	lastLoadedEvent string
+}
+
+func NewEventSystem() *eventSystem {
+	return &eventSystem{eventHandlers: make(map[reflect.Type]EventHandler, 0)}
+}
+
+func (system *eventSystem) RegisterEventHandler(eventType reflect.Type, handler EventHandler) {
+	system.eventHandlers[eventType] = handler
+}
+
+func (system eventSystem) RegisteredEventTypes() []reflect.Type {
+	types := make([]reflect.Type, 0, len(system.eventHandlers))
+	for t, _ := range system.eventHandlers {
+		types = append(types, t)
+	}
+	return types
+}
+
+func (system eventSystem) ProcessEvent(event interface{}) error {
+	eventType := reflect.TypeOf(event)
+	handler := system.eventHandlers[eventType]
+	if handler == nil {
+		return errors.New(fmt.Sprintf("no handler registered for event (type = %T), handlers are: %v", event, system.eventHandlers))
+	}
+	return handler(event)
+}
+
+func (system eventSystem) PublishEvent(event interface{}) error {
+	if reflect.TypeOf(event).Kind() == reflect.Struct {
+		// TODO: This is actually a pretty ugly API, so this should be fixed in future
+		return errors.New("a struct was passed to PublishEvent, which should've been a pointer to the struct")
+	}
+	// TODO: In future we'll publish the event to the mesh network here
+	return system.ProcessEvent(event)
+}
+
+// Load the db files from the datadir
+func (system *eventSystem) LoadAndApplyEvents(location string) error {
+	err := validateLocation(location)
+	if err != nil {
+		return err
+	}
+
+	type fileEvent struct {
+		file  string
+		event interface{}
+	}
+	events := make([]fileEvent, 0)
+	entries, err := ioutil.ReadDir(location)
+	for i := system.findStartIndex(entries); i < len(entries); i++ {
+		entry := entries[i]
+		if entry.IsDir() {
+			continue
```

Comment:
you could add a feature for recursive traversal, allowing for grouping of events when them list becomes longer and longer

Reply:
Could be a good idea, although I don't think this file-based storage will still be there at that point in time, since we might want use some sorts of simple, local, append-only database instead of rolling our own.